src/algorithm/HISTOGRAM-Cuda.cpp

```diff
@@ -0,0 +1,356 @@
+              RAJA::cuda::ConstantPreferredReplicationConcretizer<preferred_global_replication>>,
+            RAJA::cuda::warp_global_xyz<>,
+            RAJA::GetOffsetLeft<int>>>>;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      HISTOGRAM_INIT_COUNTS_RAJA(multi_reduce_policy);
+
+      RAJA::forall<RAJA::cuda_exec<block_size, true /*async*/>>( res,
+          RAJA::RangeSegment(ibegin, iend),
+          [=] __device__ (Index_type i) {
+        HISTOGRAM_BODY;
+      });
+
+      HISTOGRAM_FINALIZE_COUNTS_RAJA(multi_reduce_policy);
+
+    }
+    stopTimer();
+
+  } else {
+     getCout() << "\n  HISTOGRAM : Unknown Cuda variant id = " << vid << std::endl;
+  }
+
+}
+
+
+void HISTOGRAM::runCudaVariant(VariantID vid, size_t tune_idx)
+{
+  size_t t = 0;
+
+  if ( vid == Base_CUDA ) {
+
+    if (tune_idx == t) {
+
+      runCudaVariantLibrary(vid);
+
+    }
+
+    t += 1;
+
+  }
+
+  if ( vid == Base_CUDA || vid == RAJA_CUDA ) {
+
+    seq_for(gpu_block_sizes_type{}, [&](auto block_size) {
+
+      if (run_params.numValidGPUBlockSize() == 0u ||
+          run_params.validGPUBlockSize(block_size)) {
+
+        seq_for(cuda_atomic_global_replications_type{}, [&](auto global_replication) {
+
+          if (run_params.numValidAtomicReplication() == 0u ||
+              run_params.validAtomicReplication(global_replication)) {
+
+            seq_for(cuda_atomic_shared_replications_type{}, [&](auto shared_replication) {
+
+              if (tune_idx == t) {
+
+                setBlockSize(block_size);
+                runCudaVariantAtomicRuntime<decltype(block_size)::value,
+                                            decltype(global_replication)::value,
+                                            decltype(shared_replication)::value>(vid);
+
+              }
+
+              t += 1;
+
+            });
+
+          }
+
+        });
+
+      }
+
+    });
+
+  } else {
+
+    getCout() << "\n  HISTOGRAM : Unknown Cuda variant id = " << vid << std::endl;
+
+  }
+
+}
+
+void HISTOGRAM::setCudaTuningDefinitions(VariantID vid)
+{
+  if ( vid == Base_CUDA ) {
+
+    addVariantTuningName(vid, "cub");
+
+  }
+
+  if ( vid == Base_CUDA || vid == RAJA_CUDA ) {
+
+    seq_for(gpu_block_sizes_type{}, [&](auto block_size) {
+
+      if (run_params.numValidGPUBlockSize() == 0u ||
+          run_params.validGPUBlockSize(block_size)) {
+
+        seq_for(cuda_atomic_global_replications_type{}, [&](auto global_replication) {
+
+          if (run_params.numValidAtomicReplication() == 0u ||
+              run_params.validAtomicReplication(global_replication)) {
+
+            seq_for(cuda_atomic_shared_replications_type{}, [&](auto shared_replication) {
+
+              addVariantTuningName(vid, "atomic_shared("+std::to_string(shared_replication)+
+                                        ")_global("+std::to_string(global_replication)+
+                                        ")_"+std::to_string(block_size));
```

Comment:
I left the global and shared replication tuning parameters in here so we could try out a sweep over them later to tune for new HW, but it does end up leaving the tunings with fairly long names. It also causes me to have different default tuning params for cuda and hip. Finally to do this I am writing the RAJA multi_reduce policy from a low level and not using the high level policy so its possible for the tuning here and the tuning in RAJA to get out of sync.

Reply:
I also did not add the with_reduce policy vs normal policy option that the reduce kernels have, though I expect the with_reduce policies to be faster as they use fewer global atomics.

Reply:
I added the occupancy vs normal policy option